cf_xarray/parametric.py

```diff
@@ -0,0 +1,691 @@
+    },
+    "height_above_reference_ellipsoid": {
+        "zlev": "height_above_reference_ellipsoid",
+        "eta": "sea_surface_height_above_ reference_ellipsoid",
+        "depth": "sea_floor_depth_below_ reference_ellipsoid",
+    },
+    "height_above_mean_sea_level": {
+        "zlev": "height_above_mean_sea_level",
+        "eta": "sea_surface_height_above_mean_ sea_level",
+        "depth": "sea_floor_depth_below_mean_ sea_level",
+    },
+}
+
+
+def _derive_ocean_stdname(**kwargs):
+    """Derive standard name for computer ocean coordinates.
+
+    Uses the concatentation of formula terms `zlev`, `eta`, and `depth`
+    standard names to compare against formula term and standard names
+    from a table. This can occur with any combination e.g. `zlev`, or
+    `zlev` + `depth`. If a match is found the standard name for the
+    computed value is returned.
+
+    Parameters
+    ----------
+    zlev : dict
+        Attributes for `zlev` variable.
+    eta : dict
+        Attributes for `eta` variable.
+    depth : dict
+        Attributes for `depth` variable.
+
+    Returns
+    -------
+    str
+        Standard name for the computer value.
+
+    Raises
+    ------
+    ValueError
+        If `kwargs` is empty, missing values for `kwargs` keys, or could not derive the standard name.
+
+    References
+    ----------
+    Please refer to the CF conventions document :
+      1. https://cfconventions.org/cf-conventions/cf-conventions.html#table-computed-standard-names
+    """
+    found_stdname = None
+
+    allowed_names = {"zlev", "eta", "depth"}
+
+    if len(kwargs) == 0 or not (set(kwargs) <= allowed_names):
+        raise ValueError(
+            f"Must provide atleast one of {', '.join(sorted(allowed_names))}."
+        )
+
+    search_term = ""
+
+    for x, y in sorted(kwargs.items(), key=lambda x: x[0]):
+        try:
+            search_term = f"{search_term}{y['standard_name']}"
+        except TypeError:
+            raise ValueError(
+                f"The values for {', '.join(sorted(kwargs.keys()))} cannot be `None`."
+            ) from None
+        except KeyError:
+            raise ValueError(
+                f"The standard name for the {x!r} variable is not available."
+            ) from None
+
+    for x, y in ocean_stdname_map.items():
+        check_term = "".join(
+            [
+                y[i]
+                for i, j in sorted(kwargs.items(), key=lambda x: x[0])
+                if j is not None
+            ]
+        )
+
+        if search_term == check_term:
+            found_stdname = x
+
+            break
+
+    if found_stdname is None:
+        stdnames = ", ".join(
+            [y["standard_name"] for _, y in sorted(kwargs.items(), key=lambda x: x[0])]
+        )
+
+        raise ValueError(
+            f"Could not derive standard name from combination of {stdnames}."
+        )
+
+    return found_stdname
+
+
+def check_requirements(func, terms):
```

Comment:
This is also quite clever. But I'd think it'd be more readable if we instead wrote a dataclass per transform

```python

@dataclass
class AtmosphereLnPressureCoordinate():
    p0: int | float
    lev: xr.DataArray

    def __init__(self, *, p0, lev):
        self.p0 = p0
        self.lev = lev

    @classmethod
    def from_terms(cls, terms: dict):
        # verify terms here.
        # set properties
        return cls(p0=p0, lev=lev)

    def decode(self):
        # do calculation
        pass
```

Reply:
Now that i look at this, it would be good to define an ABC that all the individual coordinates subclass from, something like

```python
from abc import ABC, abstractmethod

class ParametricVerticalCoordinate(ABC):
    @abstractmethod
    @classmethod
    def from_terms(cls, terms: dict):
        # verify terms here.
        ...

    @abstractmethod
    def decode(self):
        ...

    @property
    @abstractmethod
    def computed_standard_name(self):
        ...
```